### PR TITLE
Support new GraphQL Header

### PR DESCRIPTION
**Pull request by supernova-at**
## Description

This PR adds a new Apollo Link to our Apollo client that reads and attaches the new `X-Magento-Cache-Id` header to every GraphQL request.

## Related Issue

Closes PWA-1868.

### Verification Stakeholders

@jimbo @pdohogne-magento 

## Test Plan

You will want to point to a backend environment that sends the `x-magento-cache-id` header.

#### Test scenario(s) for direct fix/feature

1. Load the homepage and wait for all requests to complete
2. Look for the `X-Magneto-Cache-Id` header of the last GraphQL request in the Network tab (response headers)
3. Perform an action that causes another GraphQL request (ex: click the shopping bag / cart icon)
4. Inspect the outgoing GraphQL request and verify that it has an `X-Magento-Cache-Id` request header and that the value matches the previous value sent by the server
5. Perform an operation that changes the cache id (ex: log in/out, change store, change currency, change address)
6. Verify that Venia properly updates and sends out its now-refreshed `X-Magento-Cache-Id` value

#### Test scenario(s) for any existing impacted features/areas

n/a

#### Test scenario(s) for any Magento Backend Supported Configurations

n/a

#### Is Browser/Device testing needed?

No

#### Any ad-hoc/edge case scenarios that need to be considered?

Test against environments that don't support the new header, perhaps.

## Checklist

-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.

### CODE REVIEW

File: packages/peregrine/lib/Apollo/magentoGqlCacheLink.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+            return {
+                ...previousContext,
+                headers: withCacheHeader
+            };
+        });
+
+        // Update the cache id with each response.
+        const updateCacheId = data => {
+            this.#cacheId = Math.random();
```
Comment (supernova-at) [inline packages/peregrine/lib/Apollo/magentoGqlCacheLink.js:28]: Actually read this from the response headers.

File: packages/peregrine/lib/Apollo/magentoGqlCacheLink.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+            console.log('received', this.#cacheId);
```
Comment (supernova-at) [inline packages/peregrine/lib/Apollo/magentoGqlCacheLink.js:30]: Remove the `console.log` statements.

File: packages/peregrine/lib/Apollo/magentoGqlCacheLink.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    request(operation, forward) {
+        // Attach the cache header to each outgoing request.
+        operation.setContext(previousContext => {
+            const { headers } = previousContext;
+            const withCacheHeader = {
+                ...headers,
+                [CACHE_ID_HEADER]: this.#cacheId
+            };
+
+            console.log('sending out', this.#cacheId);
```
Comment (supernova-at) [inline packages/peregrine/lib/Apollo/magentoGqlCacheLink.js:18]: Remove the `console.log` statements.

File: packages/peregrine/lib/Apollo/magentoGqlCacheLink.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+const storage = new BrowserPersistence();
+
+/**
+ * The Magento GraphQL Cache Link class is an ApolloLink that is responsible for
+ * Venia sending the proper `x-magento-cache-id` header with each of its GraphQL requests.
+ */
+export default class MagentoGQLCacheLink extends ApolloLink {
+    // The links get reinstantiated on refresh.
+    // If we have an existing cache id value from a previous browsing session, use it.
+    #cacheId = storage.getItem(LOCAL_STORAGE_KEY) || null;
```
Comment (pdohogne-magento) [inline packages/peregrine/lib/Apollo/magentoGqlCacheLink.js:17]: Question: When does this get instantiated to re-check local storage?

File: packages/peregrine/lib/Apollo/__tests__/magentoGqlCacheLink.spec.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    // Act.
+    instance.request(operation, forward);
+
+    // Assert.
+    expect(mockSetItem).toHaveBeenCalledTimes(1);
+    expect(mockSetItem).toHaveBeenCalledWith('magento_cache_id', 'unit_test');
+});
+
+test('it does not update its internal value when not present in the response', () => {
```
Comment (pdohogne-magento) [inline packages/peregrine/lib/Apollo/__tests__/magentoGqlCacheLink.spec.js:100]: Is there a reason for this?

**Comment by fooman** [comment]
I have been working hard on our project to back out of what I consider a wrong direction; adding custom http headers to the graphQL requests. 

I haven't found the time yet to write this up as an issue report but the summary of it is that CORS will kick in (essentially doubling network activity) if a request is more than Simple (when multiple domains are involved).  Details [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). This pull request seems to be doing more of the same when we should try to do the opposite (we should keep it a simple request if we can). If this is driven by requirements of fastly this should get implemented as a separate optional package.

In my view a better suited solution would be https://github.com/danslo/magento2-module-automatic-persisted-queries.

**Review by jimbo — APPROVED** [review]
@supernova-at Thanks Andy, this is exactly what I had in mind.

Older versions of PWA Studio will have to create and apply this link elsewhere, since `useAdapter` won't exist for them, but it should be easy to drop in.

**Comment by pdohogne-magento** [comment]
@jimbo How old of a version?  We have a client on Magento 2.4.2 who needs this ASAP, but I don't know what version of PWA they're on

File: packages/peregrine/lib/Apollo/magentoGqlCacheLink.js (hunk 4 above)
Comment (supernova-at) [reply packages/peregrine/lib/Apollo/magentoGqlCacheLink.js:17]: >When does this get instantiated to re-check local storage?

Every time the app is loaded: both on initial navigation to the site and on a browser refresh.

> And when does local storage get cleaned (if ever)?

I did not put any expiration date on this piece of data in local storage, it will stick around forever.

**Comment by supernova-at** [comment]
>How old of a version? We have a client on Magento 2.4.2 who needs this ASAP, but I don't know what version of PWA they're on

@pdohogne-magento Version 11 will introduce `useAdapter`. This change will likely be released in an `11.1.0`. They'll just have to modify the `Adapter` component instead of `useAdapter`.

**Comment by anthoula** [comment]
@supernova-at Please take a look at this new lighthouse warning. Wondering if it's relevant on my local instance. 
`Avoid enormous network payloads Total size was 3,185 KiB`

Screen Shot 2021-07-21 at 3 32 49 PM

**Comment by anthoula** [comment]
✅   QA Approved

QA Verification Steps [PWA-1868](https://jira.corp.magento.com/browse/PWA-1868)
